docs/posts/llms-and-data/index.qmd

```diff
@@ -0,0 +1,258 @@
+
+load_dotenv()  # <2>
+
+con = ibis.connect("duckdb://penguins.ddb")  # <3>
+t = ibis.examples.penguins.fetch()  # <3>
+t = con.create_table("penguins", t.to_pyarrow(), overwrite=True)  # <3>
+```
+
+1. Import the libraries we need.
+2. Load the environment variable to setup Marvin to call our OpenAI account.
+3. Setup the demo datain an Ibis backend.
+
+```{python}
+import ibis  # <1>
+import marvin  # <1>
+
+from ibis.expr.schema import Schema  # <1>
+from ibis.expr.types.relations import Table  # <1>
+
+ibis.options.interactive = True  # <2>
+marvin.settings.llm_model = "openai/gpt-4"  # <2>
+
+con = ibis.connect("duckdb://penguins.ddb")  # <3>
+t = con.table("penguins")  # <3>
+```
+
+1. Import Ibis and Marvin.
+2. Configure Ibis (interactive) and Marvin (GPT-4).
+3. Connect to the data and load a table into a variable.
+
+```{python}
+@marvin.ai_fn  # <1>
+def _generate_sql_select(
+    text: str, table_name: str, table_schema: Schema
+) -> str:  # <1>
+    """Generate SQL SELECT from text."""  # <1>
+
+
+def sql_from_text(text: str, t: Table) -> Table:  # <2>
+    """Run SQL from text."""  # <2>
+    return t.sql(_generate_sql_select(text, t.get_name(), t.schema()).strip(";"))  # <2>
+```
+
+1. A non-deterministic, LLM-powered AI function.
+2. A deterministic, human-authored function that calls the AI function.
+
+```{python}
+t2 = sql_from_text("the unique combination of species and islands", t)
+t2
+```
+
+```{python}
+t3 = sql_from_text(
+    "the unique combination of species and islands, with their counts, ordered from highest to lowest, and name that column just 'count'",
+    t,
+)
+t3
+```
+
+This works well-enough for simple cases and can be expanded to handle complex
+ones. In many scenarios, it may be easier to express a query in English or
+another language than to write it in SQL, especially if working across multiple
+SQL dialects.
+
+### Approach 2: LLM writes a subroutine
+
+If more complex logic needs to be expressed, SoTA LLMs are also decent at
+writing Python and a number of other programming languages. Many data platforms
+support user-defined functions (UDFs) in Python or some other language. We'll
+stick to scalar Python UDFs via DuckDB to demonstrate the concept:
+
+```{python}
+@marvin.ai_fn  # <1>
+def _generate_python_function(text: str) -> str:  # <1>
+    """Generate a simple, typed, correct Python function from text."""  # <1>
+
+
+def create_udf_from_text(text: str) -> str:  # <2>
+    """Create a UDF from text."""  # <2>
+    return f"""
+import ibis
+
+@ibis.udf.scalar.python
+{_generate_python_function(text)}
+""".strip()  # <2>
+```
+
+1. A non-deterministic, LLM-powered AI function.
+2. A deterministic, human-authored function that calls the AI function.
+
+```{python}
+udf = create_udf_from_text(
+    "a function named count_vowels that given an input string, returns an int w/ the number of vowels (y_included as a boolean option defaulted to False)"
+)
+print(udf)
+exec(udf)
```

Comment:
put the exec inside the `create_udf_from_text` function. Also for due diligence I would mention how this is a total security nightmare if you are passing in untrusted user input.

Reply:
why move it inside the function? with it outside, you (directly or via the bot this will be called from) can inspect the function before you execute/register it

fyi this post has moved over here: https://ibis-project.github.io/ibis-birdbrain/posts/llms-and-data-pt1/

I'm going to have a series of posts to start off that new project, and will make one here in Ibis to announce. don't want to overwhelm the Ibis blog or start making it all about AI/LLMs

Reply:
Eh, just a bit more encapsulated, but yes that is a nice benefit that is probably worth it. Sure keep it outside. Sounds good